### PR TITLE
`sort` for NTuple and other iterables

### LINKED ISSUE #803: `sort` for `NTuple` and other iterables
`sort` for `NTuple` and other iterables was added in https://github.com/JuliaLang/julia/pull/46104, it would be nice to add that to `Compat`.

### PR DESCRIPTION
**Pull request by mtfishman**
Closes #803.

### CODE REVIEW

File: src/Compat.jl [hunk 1]
```diff
@@ -706,6 +706,42 @@ if VERSION < v"1.7.0-DEV.1187"
     export redirect_stdio
 end
 
+# https://github.com/JuliaLang/julia/pull/46104
+if VERSION < v"1.10.0-DEV.1404"
+    import Base: sort, Ordering, Forward, ord, lt, tail, copymutable, DEFAULT_STABLE, IteratorSize, HasShape, IsInfinite
```
Comment (martinholters) [inline src/Compat.jl:711]: Shouldn't import things not meant to be extended:
```suggestion
    import Base: sort
    using Base: Ordering, Forward, ord, lt, tail, copymutable, DEFAULT_STABLE, IteratorSize, HasShape, IsInfinite
```
or (and that would be my personal preference)
```suggestion
    using Base: Ordering, Forward, ord, lt, tail, copymutable, DEFAULT_STABLE, IteratorSize, HasShape, IsInfinite
```
and explicitly qualify `sort` below.

**Review by martinholters — COMMENTED** [review]
LGTM except for some minor nit regarding `import` usage.

File: src/Compat.jl [hunk 2]
```diff
@@ -706,6 +705,43 @@ if VERSION < v"1.7.0-DEV.1187"
...
+# https://github.com/JuliaLang/julia/pull/46104
+if VERSION < v"1.10.0-DEV.1404"
+    using Base: Ordering, Forward, ord, lt, tail, copymutable, DEFAULT_STABLE, IteratorSize, HasShape, IsInfinite
+    function Base.sort(v; kws...)
+        size = IteratorSize(v)
+        size == HasShape{0}() && throw(ArgumentError("$v cannot be sorted"))
+        size == IsInfinite() && throw(ArgumentError("infinite iterator $v cannot be sorted"))
+        sort!(copymutable(v); kws...)
+    end
+    sort(v::AbstractVector; kws...) = sort!(copymutable(v); kws...) # for method disambiguation
```
Comment (martinholters) [inline src/Compat.jl:717]: I think this line just moved in https://github.com/JuliaLang/julia/issues/46104 and so there is no need to reproduce it here?
```suggestion
```
AFAICT, it is present in all relevant past versions:
https://github.com/JuliaLang/julia/blob/v1.6.0/base/sort.jl#L794
https://github.com/JuliaLang/julia/blob/v1.7.0/base/sort.jl#L770
https://github.com/JuliaLang/julia/blob/v1.8.0/base/sort.jl#L768
https://github.com/JuliaLang/julia/blob/v1.9.0/base/sort.jl#L1400

Is there a specific reason you added?

Comment (mtfishman) [reply src/Compat.jl:717]: Ah right. I knew there was a reason I left it out to begin with.

Comment (martinholters) [reply src/Compat.jl]: Doing the first review round, I had wondered why this line was missing until I found it had just moved...

**Comment by ericphanson** [comment]
This should probably be reverted as it was reverted upstream: https://github.com/JuliaLang/julia/pull/52010

**Comment by martinholters** [comment]
#811

**Pull request merged**